File: matrix/sol.java
package matrix;

import java.util.Arrays;
import java.util.Scanner;

/**
 * sol
 */
public class sol {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String targetWord = scanner.nextLine().trim();
        int[] dimensions = Arrays.stream(scanner.nextLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        String[] matrix = new String[dimensions[0]];

        for (int i = 0; i < dimensions[0]; i++)
            matrix[i] = scanner.nextLine();

        System.out.println(solution(targetWord, dimensions, matrix));

        scanner.close();
    }

    public static String solution(String targetWord, int[] dimensions, String[] matrix) {
        // init
        int rows = dimensions[0];
        int cols = dimensions[1];

        // algo
        for (int i = 0; i < rows; i++) {
            if (checkHorizontal(matrix, targetWord, i)) {
              return String.format("row %d", i + 1);
            }
          }
        
          for (int i = 0; i < cols; i++) {
            if (checkVertical(matrix, targetWord, i)) {
              return String.format("col %d", i + 1);
            }
          }

        return "none";
    }

    public static String getColumn(String[] matrix, int colNumber) {
        String columnStr = "";
        for (String row : matrix) {
            columnStr = columnStr + row.charAt(colNumber);
        }
        return columnStr;
      }
      
      // helper functions
      public static Boolean checkHorizontal(String[] matrix, String targetWord, int row) {
        String right = matrix[row].repeat(2);
        String left = (new StringBuffer(matrix[row]).reverse().toString()).repeat(2);
      
        return right.indexOf(targetWord) != -1 || left.indexOf(targetWord) != -1;
      }
      
      public static Boolean checkVertical(String[] matrix, String targetWord, int col) {
        String columnStr = getColumn(matrix, col);
        String bottom = columnStr.repeat(2);
        String top = (new StringBuffer(columnStr).reverse().toString()).repeat(2);
      
        return bottom.indexOf(targetWord) != -1 || top.indexOf(targetWord) != -1;
      }
}

File: matrix/sol.js
const { cp } = require("fs");

const readline = require("readline").createInterface({ input: process.stdin });

const input = {
  matrix: [],
  str: "",
};

let index = 0;
let dimensions = [];

readline.on("line", (line) => {
  if (index == 0) {
    input.str = line;
    index++;
    return;
  }
  if (index == 1) {
    dimensions = line.split(" ").map((val) => parseInt(val));
    index++;
    return;
}
if (index < 2 + dimensions[0]) {
    input.matrix.push(line);
    index++;
    if (index !== 2 + dimensions[0]) return
  }
  readline.close();
  const { matrix, str } = input;
  const rows = matrix.length;
  const cols = matrix[0].length;

  for (let i = 0; i < rows; i++) {
    if (checkHorizontal(matrix, str, i)) {
      console.log(`row ${i + 1}`);
      return;
    }
  }

  for (let i = 0; i < cols; i++) {
    if (checkVertical(matrix, str, i)) {
      console.log(`col ${i + 1}`);
      return;
    }
  }

  console.log("none");
});
// utility functions
function reverseString(str) {
  return str.split("").reverse().join("");
}

function getColumn(matrix, colNumber) {
  return matrix.map((row) => row[colNumber]).join("");
}

// helper functions
function checkHorizontal(matrix, str, pos) {
  const row = Math.floor(pos / matrix.length);
  const right = matrix[row].repeat(2);
  const left = reverseString(matrix[row]).repeat(2);

  return right.search(str) != -1 || left.search(str) != -1;
}

function checkVertical(matrix, str, pos) {
  const col = pos % matrix[0].length;
  const bottom = getColumn(matrix, col).repeat(2);
  const up = reverseString(getColumn(matrix, col)).repeat(2);

  return bottom.search(str) != -1 || up.search(str) != -1;
}
